File: src/shamir/restore_secret.c
#include <stdlib.h>
#include <string.h>

#include "shamir.h"

int
restore_secret(const unsigned int SHARES_LEN, const unsigned int SECRET_LEN,
               const uint8_t shares[SHARES_LEN][SHARE_LEN(SECRET_LEN)],
               uint8_t secret[SECRET_LEN])
{
  size_t i, j;

  if (SHARES_LEN < 2) return -1;
  if (SHARES_LEN > FIELD - 1) return -2;

  uint8_t(*points)[2] = malloc(sizeof(uint8_t[SHARES_LEN][2]));
  if (points == NULL) return -3;

  for (i = 0; i < SECRET_LEN; i++)
  {
    for (j = 0; j < SHARES_LEN; j++)
    {
      memcpy(&points[j][0], &shares[j][SECRET_LEN], sizeof(uint8_t));
      memcpy(&points[j][1], &shares[j][i], sizeof(uint8_t));
    }

    secret[i] = interpolate(SHARES_LEN, points);
  }

  free(points);

  return 0;
}

File: src/shamir/restoreSecret.ts
import demosMemory from "./memory";

import libdemos from "@libdemos";

import type { LibDemos } from "@libdemos";

const restoreSecret = async (shares: Uint8Array[], module?: LibDemos) => {
  const sharesLen = shares.length;

  const shareItemLen = shares[0].length;
  const lengthVerification = shares.every((v) => v.length === shareItemLen);
  if (!lengthVerification) throw new Error("Shares length varies.");

  const secretLen = shareItemLen - 1;

  const wasmMemory =
    module?.wasmMemory ?? demosMemory.restoreSecretMemory(secretLen, sharesLen);

  const demosModule = module ?? (await libdemos({ wasmMemory })); // await shamirMethodsModule({ wasmMemory });

  const ptr1 = demosModule._malloc(
    sharesLen * (secretLen + 1) * Uint8Array.BYTES_PER_ELEMENT,
  );
  const sharesArray = new Uint8Array(
    wasmMemory.buffer,
    ptr1,
    sharesLen * (secretLen + 1) * Uint8Array.BYTES_PER_ELEMENT,
  );
  for (let i = 0; i < sharesLen; i++) {
    sharesArray.set(shares[i], i * (secretLen + 1));
  }

  const ptr2 = demosModule._malloc(secretLen * Uint8Array.BYTES_PER_ELEMENT);
  const secretArray = new Uint8Array(
    wasmMemory.buffer,
    ptr2,
    secretLen * Uint8Array.BYTES_PER_ELEMENT,
  );

  const result = demosModule._restore_secret(
    sharesLen,
    secretLen,
    sharesArray.byteOffset,
    secretArray.byteOffset,
  );

  demosModule._free(ptr1);

  switch (result) {
    case 0: {
      const sec = Uint8Array.from(secretArray);
      demosModule._free(ptr2);

      return sec;
    }

    case -1: {
      demosModule._free(ptr2);

      throw new Error("Not enough shares provided.");
    }

    case -2: {
      demosModule._free(ptr2);

      throw new Error("Need at most 255 shares.");
    }

    default: {
      demosModule._free(ptr2);

      throw new Error("An unexpected error occured.");
    }
  }
};

export default restoreSecret;
